nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RRestoreFacetImpl.java

```diff
@@ -0,0 +1,94 @@
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.r.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+
+import javax.inject.Named;
+
+import org.sonatype.nexus.common.collect.AttributesMap;
+import org.sonatype.nexus.repository.FacetSupport;
+import org.sonatype.nexus.repository.r.RFacet;
+import org.sonatype.nexus.repository.r.RRestoreFacet;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.storage.AssetBlob;
+import org.sonatype.nexus.repository.storage.Component;
+import org.sonatype.nexus.repository.storage.Query;
+import org.sonatype.nexus.repository.storage.StorageTx;
+import org.sonatype.nexus.repository.transaction.TransactionalTouchBlob;
+import org.sonatype.nexus.repository.view.Content;
+import org.sonatype.nexus.transaction.UnitOfWork;
+
+import static org.sonatype.nexus.repository.r.internal.RDescriptionUtils.extractDescriptionFromArchive;
+import static org.sonatype.nexus.repository.storage.ComponentEntityAdapter.P_VERSION;
+import static org.sonatype.nexus.repository.storage.MetadataNodeEntityAdapter.P_NAME;
+
+/**
+ * @since 1.1.next
+ */
+@Named
+public class RRestoreFacetImpl
+    extends FacetSupport
+    implements RRestoreFacet
+{
+  @Override
+  @TransactionalTouchBlob
+  public void restore(final AssetBlob assetBlob, final String path) throws IOException {
+    StorageTx tx = UnitOfWork.currentTx();
+    RFacet facet = facet(RFacet.class);
+
+    Asset asset;
+    if (componentRequired(path)) {
+      Map<String, String> attributes;
+
+      try (InputStream is = assetBlob.getBlob().getInputStream()) {
+        attributes = extractDescriptionFromArchive(path, is);
+      }
+
+      Component component = facet.findOrCreateComponent(tx, attributes);
+      asset = facet.findOrCreateAsset(tx, component, path, attributes);
+    }
+    else {
+      asset = facet.findOrCreateAsset(tx, path);
+    }
+    tx.attachBlob(asset, assetBlob);
+
+    Content.applyToAsset(asset, Content.maintainLastModified(asset, new AttributesMap()));
+    tx.saveAsset(asset);
+  }
+
+  @Override
+  @TransactionalTouchBlob
+  public boolean assetExists(final String path) {
+    final StorageTx tx = UnitOfWork.currentTx();
+    return tx.findAssetWithProperty(P_NAME, path, tx.findBucket(getRepository())) != null;
+  }
+
+  @Override
+  public boolean componentRequired(final String name) {
+    return !(name.contains("PACKAGES") || name.endsWith(".rds"));
```

Comment:
Question: Is it possible to use absolute string check here for `PACKAGES`. What would happen if an R package called `Util_PACKAGES` was available? Is it possible to makes use of `/` to narrow down the change of error

Reply:
There are no archives with names that included uppercase PACKAGES. But nobody forbids to add this archive.

Reply:
done 2e3d35497ffd3a4947c8e4d441eeba9e2d52b0ba